### PR TITLE
Issue/8666 quick start reminder notifications

### LINKED ISSUE #8666: Quick Start: Add Reminder Notification

### PR DESCRIPTION
**Pull request by khaykov**
Fixes #8666

This PR adds quick start reminder notifications.
To avoid overcomplicating logic, I had to balance requirements and implementation a bit.

Current notification logic is as follows:
- Reminder for the next uncomplete task is scheduled to go off in 48 hours after completing a task (which means reminder automatically starts after you stat QS and complete Create Site task).
- Completing a task cancels current reminder and schedules a new one.
- Next task for a reminder is picked from the same type of tasks as a completed task. If no uncomplete tasks of the type left the uncomplete task from the other task type will be picked.
- A reminder will wake the device with notification.
- Clicking on the notification will open app even if it was closed, navigate to MySite tab and start selected task (pointer + snackbar).
- A reminder will not be scheduled when you complete Edit Post task before starting QS.
- A reminder will be canceled on logout.
- A reminder will be canceled on new site creation.
- A reminder will be canceled when you switch active site.
- A reminder will be canceled if the device reboots.

@SylvesterWilmott Can you confirm if the logic is ok?

The abovementioned logic is also the test items. I suggest reducing the delay for notification to something short [here](https://github.com/wordpress-mobile/WordPress-Android/blob/cadd0953236e314e3111a14ff815abe871c9ec3f/WordPress/src/main/java/org/wordpress/android/util/QuickStartUtils.kt#L46).

I don't have much experience working with Alarm Manager, Broadcast receivers and notifications, so If I missed something or something could be done in a better way please let me know :)

### CODE REVIEW

**Comment by SylvesterWilmott** [comment]
>Clicking on the notification will open app even if it was closed, navigate to MySite tab and start selected task (pointer + snackbar).

I think we should simply open the app at the site dashboard, it might not be clear what is happening if the first task that was completed wasn't explicitly through our list.

Other than that it looks good to me!

**Comment by theck13** [comment]
@khaykov, do we need to make changes after @SylvesterWilmott's [comment](https://github.com/wordpress-mobile/WordPress-Android/pull/8902#issuecomment-456491995)?  If not, let me know and I'll start the review.

**Comment by khaykov** [comment]
@SylvesterWilmott  I changed the behavior so the QS task will not be started 👍 

@theck13 Thanks for waiting! I think this PR is ready for review :)

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java [hunk 1]
```diff
@@ -734,6 +741,13 @@ public void onClick(View v) {
                 if (getMySiteFragment() != null) {
                     getMySiteFragment().onActivityResult(requestCode, resultCode, data);
 
+                    boolean isSameSiteSelected = data != null
```
Comment (theck13) [inline WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Is the `data != null` null check necessary?  It usually makes sense to do before something like `data.getIntExtra`, but removing it [here](https://github.com/wordpress-mobile/WordPress-Android/blob/d60a2e2ac26f570613c91e6f85aaa86c04ad6ab0/WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java#L744) and [here](https://github.com/wordpress-mobile/WordPress-Android/blob/d60a2e2ac26f570613c91e6f85aaa86c04ad6ab0/WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java#L754) doesn't produce lint warnings.

Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: My reasoning behind the check is that the `setSite(data)` and `jumpNewPost(data)` that are used right after that do check if the data is null, so I assumed that it actually might be`null`, which could lead to NPE.

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        Intent resultIntent = new Intent(context, WPMainActivity.class);
+        resultIntent.putExtra(MySiteFragment.ARG_QUICK_START_TASK, quickStartTaskDetails.getTask());
+        resultIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK
+                              | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+        resultIntent.setAction(Intent.ACTION_MAIN);
+        resultIntent.addCategory(Intent.CATEGORY_LAUNCHER);
+        PendingIntent notificationContentIntent = PendingIntent.getActivity(context, QUICK_START_REMINDER_NOTIFICATION,
+                resultIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+
+        NotificationManager notificationManager = (NotificationManager) context.getSystemService(NOTIFICATION_SERVICE);
```
[the remark this answers is not in the document]
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java]: Thanks, didn't know about it!

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        Notification notification = new NotificationCompat.Builder(context,
+                context.getString(R.string.notification_channel_normal_id))
+                .setSmallIcon(R.drawable.ic_my_sites_24dp)
+                .setContentTitle(context.getString(quickStartTaskDetails.getTitleResId()))
+                .setContentText(context.getString(quickStartTaskDetails.getSubtitleResId()))
+                .setOnlyAlertOnce(true)
+                .setAutoCancel(true)
+                .setContentIntent(notificationContentIntent)
+                .build();
```
[the remark this answers is not in the document]
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java]: If I understand this correctly, we will also need to have a service that runs in the background and listens to this `PendingIntent`. If this is correct, I wonder, do we need it just for this one task?

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        resultIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK
+                              | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+        resultIntent.setAction(Intent.ACTION_MAIN);
+        resultIntent.addCategory(Intent.CATEGORY_LAUNCHER);
+        PendingIntent notificationContentIntent = PendingIntent.getActivity(context, QUICK_START_REMINDER_NOTIFICATION,
+                resultIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+
+        NotificationManager notificationManager = (NotificationManager) context.getSystemService(NOTIFICATION_SERVICE);
+        Notification notification = new NotificationCompat.Builder(context,
+                context.getString(R.string.notification_channel_normal_id))
```
[the remark this answers is not in the document]
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java]: Thanks for the explanation! I was not too sure how channels work, but it makes sense now.

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java]: We can use a `BroadcastReceiver` to catch when the notification is dismissed with something like this.

```java
public class NotificationBroadcastReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        NotificationType type = intent.getExtras().getSerializableExtra(EXTRA_NOTIFICATION_DISMISS);
        // Do something based on type.
    }
}
``` 

We can create the `PendingIntent` like this.

```java
Intent intent = new Intent(context, NotificationBroadcastReceiver.class);
intent.putExtra(EXTRA_NOTIFICATION_DISMISS, notificationType);
return PendingIntent.getBroadcast(context, RequestCodes.NOTIFICATION_BROADCAST_RECEIVER, intent, 0);
```

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java (hunk 1 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Removing the null checks from `setSite(data)` and `jumpNewPost(data)` doesn't produce any lint warnings either.  I'm not sure where the first `NullPointerException` would occur if `data` was `null`, but I guess it doesn't hurt to have those null checks.

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartReminderReceiver.java]: Looks like we already have some code in place to handle this in `NotificationsProcessingService` . I added the `PendingIntent` and TODO for tracking in this https://github.com/wordpress-mobile/WordPress-Android/pull/8902/commits/cdb854f2bc6825dfba82b3b87f39b4c3af241de8

**Comment by khaykov** [comment]
@theck13 Thanks for your patience and insight into notifications 🙇 I added a `Reminder` channel and detection of QS Notification dismissal. Should be working as expected now :)

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java [hunk 5]
```diff
@@ -225,6 +225,10 @@ public void run() {
                 boolean openedFromShortcut = (getIntent() != null && getIntent().getStringExtra(
                         ShortcutsNavigator.ACTION_OPEN_SHORTCUT) != null);
                 boolean openRequestedPage = (getIntent() != null && getIntent().hasExtra(ARG_OPEN_PAGE));
+                boolean isQuickStartRequestedFromPush = (getIntent() != null
+                                                         && getIntent()
+                                                                 .getBooleanExtra(MySiteFragment.ARG_QUICK_START_TASK,
+                                                                         false));
```
[the remark this answers is not in the document]
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Fixed 👍

**Pull request merged**